File: CSharp/src/Runtime/Scripts/XClient/interceptor.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Grpc.Core.Interceptors;

namespace FrameworkEase.Universal.X.gRPC
{
    public partial class XClient
    {
        internal class InterceptorEntry : IComparable<InterceptorEntry>
        {
            internal Interceptor Interceptor { get; }
            internal int Priority { get; }
            internal int Increment { get; }

            internal InterceptorEntry(Interceptor interceptor, int priority = 0, int increment = 0)
            {
                Interceptor = interceptor;
                Priority = priority;
                Increment = increment;
            }

            public int CompareTo(InterceptorEntry other)
            {
                if (other == null) return 1;
                var priorityComparison = Priority.CompareTo(other.Priority);
                if (priorityComparison != 0) return priorityComparison;
                return Increment.CompareTo(other.Increment);
            }
        }

        internal static readonly List<InterceptorEntry> Interceptors = new();
        internal static int InterceptorIncrement = 0;

        public static bool Intercept(Type interceptorType, int priority = 0)
        {
            if (interceptorType == null) throw new ArgumentNullException(nameof(interceptorType));
            if (!interceptorType.IsSubclassOf(typeof(Interceptor))) throw new Exception($"Interceptor type {interceptorType} is not a subclass of {typeof(Interceptor)}");
            lock (Interceptors)
            {
                if (Interceptors.Any(e => e.Interceptor.GetType() == interceptorType)) return false;
                var increment = Interlocked.Increment(ref InterceptorIncrement);
                var entry = new InterceptorEntry((Interceptor)Activator.CreateInstance(interceptorType)!, priority, increment);
                Interceptors.Add(entry);
                Interceptors.Sort();
                return true;
            }
        }
    }

    public partial class XClient<TService>
    {
        internal static readonly List<XClient.InterceptorEntry> Interceptors = new();

        public static bool Intercept(Type interceptorType, int priority = 0)
        {
            if (interceptorType == null) throw new ArgumentNullException(nameof(interceptorType));
            if (!interceptorType.IsSubclassOf(typeof(Interceptor))) throw new Exception($"Interceptor type {interceptorType} is not a subclass of {typeof(Interceptor)}");
            if (XClient.Interceptors.Any(e => e.Interceptor.GetType() == interceptorType)) return false;
            lock (Interceptors)
            {
                if (Interceptors.Any(e => e.Interceptor.GetType() == interceptorType)) return false;
                var increment = Interlocked.Increment(ref XClient.InterceptorIncrement);
                var entry = new XClient.InterceptorEntry((Interceptor)Activator.CreateInstance(interceptorType)!, priority, increment);
                Interceptors.Add(entry);
                Interceptors.Sort();
                return true;
            }
        }
    }
}

File: Go/XClient/interceptor.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XClient

import (
	"reflect"
	"sort"
	"sync"
	"sync/atomic"

	"google.golang.org/grpc"
)

type interceptorEntry struct {
	Interceptor grpc.UnaryClientInterceptor
	Priority    int
	Increment   int64
}

type interceptorList []*interceptorEntry

func (l interceptorList) Len() int { return len(l) }

func (l interceptorList) Less(i, j int) bool {
	if l[i].Priority != l[j].Priority {
		return l[i].Priority < l[j].Priority
	}
	return l[i].Increment < l[j].Increment
}

func (l interceptorList) Swap(i, j int) { l[i], l[j] = l[j], l[i] }

var (
	interceptors         = make(map[reflect.Type]interceptorList)
	interceptorsMu       sync.RWMutex
	interceptorIncrement int64
)

func Intercept[T any](interceptor grpc.UnaryClientInterceptor, priority ...int) bool {
	if interceptor == nil {
		panic("XClient.Intercept: interceptor cannot be nil.")
	}
	interceptorPtr := reflect.ValueOf(interceptor).Pointer()
	typ := reflect.TypeFor[T]()
	anyType := reflect.TypeFor[any]()

	interceptorsMu.Lock()
	defer interceptorsMu.Unlock()

	if typ != anyType {
		anyInterceptors := interceptors[anyType]
		for _, entry := range anyInterceptors {
			if reflect.ValueOf(entry.Interceptor).Pointer() == interceptorPtr {
				return false
			}
		}
	}

	typedInterceptors := interceptors[typ]
	for _, entry := range typedInterceptors {
		if reflect.ValueOf(entry.Interceptor).Pointer() == interceptorPtr {
			return false
		}
	}

	increment := atomic.AddInt64(&interceptorIncrement, 1)
	priorityValue := 0
	if len(priority) > 0 {
		priorityValue = priority[0]
	}
	entry := &interceptorEntry{
		Interceptor: interceptor,
		Priority:    priorityValue,
		Increment:   increment,
	}
	typedInterceptors = append(typedInterceptors, entry)
	sort.Sort(typedInterceptors)
	interceptors[typ] = typedInterceptors
	return true
}
